include/uapi/linux/virtio_ids.h

```diff
@@ -46,5 +46,6 @@
 #define VIRTIO_ID_IOMMU        23 /* virtio IOMMU */
 #define VIRTIO_ID_FS           26 /* virtio filesystem */
 #define VIRTIO_ID_PMEM         27 /* virtio pmem */
+#define VIRTIO_ID_ADSP         28 /* virtio AudioDSP */
```

Comment:
I'm going to propose a change here that will require a little rework, but it will be required due to multi client support and that we are no longer audio only. @plbossart fyi.

Can we use the VIRTIO_ID_RPMSG id now and use RPMSG over virtio/vhost as the transport mechanism/container for the IPC. Advantages:

1) No need for new OASIS standard.
2) Can upstream once it's working due to 1.
3) Supports the definition of a "remote processor", which can be any DSP feature and not just audio.
4) Can support external low level infrastructure and standards i.e. debug, trace.

This will need an adapter to insert IPC into RPMSG infrastructure/containers and vice versa, but the advantages are obvious.

Btw, there is an Audio ID going through OASIS atm, we should also be able to hook into that too as future work to support "legacy" audio. i.e. PCM -> volume -> DAI topology only.

Reply:
@lgirdwood ok, can do. But can we do that after we merge this? You know that GH isn't a perfect tool for reviewing multiple revisions. This change will modify both the host and the guest. So modifying the current patch stack and resubmitting it will make development, testing, and review more difficult both for me and for reviewers. Can we merge this, then I make the change and we squash them when pushing upstream?

Reply:
@lgirdwood @plbossart can we get some resolution on this, please? I could start implementing RPMSG but it would be good to agree on the process first: can we first merge, then switch, or shall I switch and we review everything again?

Reply:
Sorry for the late feedback I didn't see this proposal.

I am not very hot on a change of directions. RPMSG will for sure create new problems, e.g. it depends on 'remote proc' for boot but I am not sure the latter supports all our firmware download/D0ix cases. It'd also like to check the licenses for RPMSG. And btw if we rely on RPMSG, then it should also be done for the non-virtualized case.

So I would prefer it if we added VIRTIO support as defined now, and keep it maintained and alive in the SOF tree - just like we just did for SoundWire. We can defer the upstream part to a point where we feel comfortable doing so. In a second phase we either update the IPC mechanism or transition to RPMSG. It's rather easy as long as @lyakh helps us maintain the topic/sof-dev-rebase so that the relevant patches are squashed in the right place. And we clearly mark all virtio stuff as 'development only'.

In the mean time we use the existing solution to iron out firmware dependencies, race conditions, power transitions, and start setting up CI for this sort of solution. we need more users of this feature, it has to leave the developer-tested more and become shared by a wider SOF set of users.

Reply:
Ack, I'm fine by this process. RPMSG is probably the path of least resistance for upstreaming the virtio parts.

Reply:
RPMSG in Linux is built on top of VirtIO, so, it seems it should be possible to use it with KVM. However I don't see any VirtIO drivers in the mainline doing that. So far I see uses for asymmetric multi-processors, where Linux is running on some cores, and other OSes run on other cores. Linux is then used to load firmware (so far only in EFI format) to those other cores. Then RPMSG on top of VirtIO is used to communicate with those remote cores and their OSes. So far I don't see any examples of RPMSG used with Linux running on remote cores too. @lgirdwood do you know any such examples?

Reply:
@lyakh https://www.kernel.org/doc/Documentation/rpmsg.txt
```c 
 int rpmsg_send(struct rpmsg_channel *rpdev, void *data, int len);
```
Looks right to me, the void * means that data format is not subject to the OASIS standards and would be correct for IPC. The channel would be a mapping to the target component/module on the DSP (but we dont use this today, we just send all messages and let FW select the correct recipient).
@ranj063 fyi, this will sit on top of multi-client (and even use the RPMSG infrastructure in the future - e.g. to support cross CPU debug, trace, and other generic RPMSG features).

Reply:
@lyakh https://www.kernel.org/doc/Documentation/rpmsg.txt
```c 
 int rpmsg_send(struct rpmsg_channel *rpdev, void *data, int len);
```
Looks right to me, the void * means that data format is not subject to the OASIS standards and would be correct for IPC. The channel would be a mapping to the target component/module on the DSP (but we dont use this today, we just send all messages and let FW select the correct recipient).
@ranj063 fyi, this will sit on top of multi-client (and even use the RPMSG infrastructure in the future - e.g. to support cross CPU debug, trace, and other generic RPMSG features).

Reply:
@lyakh I mean to "Resolve conservation" for this discussion for patch "ASoC: SOF: add an RPMsg VirtIO DSP driver" and w.r.t. virtio_ids.h L49

Reply:
@kv2019i indeed it can